packages/type-utils/src/typeFlagUtils.ts

```diff
@@ -1,31 +1,83 @@
 import { unionTypeParts } from 'tsutils';
 import * as ts from 'typescript';
+import { getEnumValues } from './getEnum';
+
+const ANY_OR_UNKNOWN = ts.TypeFlags.Any | ts.TypeFlags.Unknown;
 
 /**
- * Gets all of the type flags in a type, iterating through unions automatically
+ * Gets all of the type flags in a type, iterating through unions automatically.
  */
-export function getTypeFlags(type: ts.Type): ts.TypeFlags {
-  let flags: ts.TypeFlags = 0;
+export function getTypeFlags(type: ts.Type): number | ts.TypeFlags {
```

Comment:
I would put forward that the correct type of the variable is probably `number` in this case.

Imagine the following code:

```ts
enum FooFlag {
  Flag1,
  Flag2,
  Flag3,
}

const myFlags = new Set<FooFlag>();
myFlags.add(FooFlag.Flag1);
myFlags.add(FooFlag.Flag2);
```

In this code snippet, our variable represents only the first and second flag, and not the third flag. It would be counter-intuitive (and wrong!) to type the type of `myFlags` as FooFlag, because it isn't equal to any single FooFlag. Rather, it is the composition of 0 or more FooFlags, so a Set is the correct way to represent this data.

In a similar way, the correct type of an integer containing one or more flags represented by an enum is NOT equal to the enum itself - that would imply that the variable can only be equal to exactly one enum, which is wrong, because it can be 2 or more enums combined. Additionally, it can also be a value representing 0 enums!

A thing that can have 0 or more bitflags is really a number. The main source of weirdness here is that it is actually a specific subset of a number, because it can't just be any number at all, it can only be a valid composition of 0 or more flags. But there's no native way to represent that to the TypeScript compiler without building yourself up a `BitFlag<T>` type. Not 100% sure what such a type would look like myself.

So, to answer your question:

> Would users of this rule now have to always explicitly indicate in that way that a type can contain any number

Yes, because otherwise you are lying about the real type of your variable, and treating the type as a comment that has no type safety. Instead, if a number is a subset of 0 or more bitflags, then write that as an actual code comment instead of stealing the type for this purpose!

Reply:
Sorry, I don't follow. What's wrong with that code snippet? `myFlags` is indeed a `Set<FooFlag>`; `FooFlag.Flag1` is an ok argument to provide to its `.add`?

Reply:
Nothing is wrong with the code snippet. The point of the code snippet is just to illustrate how I conceptualize what a bitflag is. Once you visualize that this is what a bitflag really is (a Set), then all of the rest follows naturally.

Reply:
To clarify, I typed it as `number | ts.TypeFlags` instead of just `number` as a code comment to imply that it exacts a composition of 0 or more `ts.TypeFlags`. A better change would be to actually use a real `BitFlag<T>` type, but that it outside of scope of this PR, so I'll just add a code comment for now.